File: tests/cpp/test_rpc.cpp
// test_rpc.cpp — RpcPipeClient / RpcPipeServer 統合テスト
// 仕様: spec/F002_rpc_message_id.md §10

#include <gtest/gtest.h>
#include "pipeutil/rpc_pipe_client.hpp"
#include "pipeutil/rpc_pipe_server.hpp"
#include "pipeutil/pipe_error.hpp"
#include "pipeutil/detail/frame_header.hpp"
#include <chrono>
#include <future>
#include <string>
#include <thread>
#include <vector>

using namespace pipeutil;
using namespace std::chrono_literals;

// ─── ユーティリティ ──────────────────────────────────────────────────────────

namespace {

std::string unique_pipe(const char* suffix) {
    return std::string("pipeutil_rpc_test_") + suffix;
}

// サーバーを非同期で起動するヘルパー
// on_server: (RpcPipeServer&) → void ラムダ
template <typename Fn>
std::future<void> async_rpc_server(const std::string& name, Fn on_server,
                                   std::chrono::milliseconds accept_timeout = 5000ms) {
    return std::async(std::launch::async,
        [name, accept_timeout, fn = std::move(on_server)]() mutable {
            RpcPipeServer srv{name};
            srv.listen();
            srv.accept(accept_timeout);
            fn(srv);
            srv.close();
        });
}

} // namespace

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-01: sizeof(FrameHeader) == 20
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcFrameTest, FrameHeader_Size_Is20) {
    static_assert(sizeof(detail::FrameHeader) == 20,
                  "FrameHeader must be 20 bytes (F-002 spec requirement)");
    EXPECT_EQ(sizeof(detail::FrameHeader), 20u);
}

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-02: 単発ラウンドトリップ
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcTest, SendRequest_SingleRoundTrip) {
    const auto pipe_name = unique_pipe("single_rt");

    auto srv_fut = async_rpc_server(pipe_name, [](RpcPipeServer& srv) {
        // serve_requests: フォアグラウンドで 1 リクエストを処理して終了
        int handled = 0;
        srv.serve_requests([&handled](const Message& req) -> Message {
            ++handled;
            std::string resp_str = "echo:" + std::string{req.as_string_view()};
            return Message{std::string_view{resp_str}};
        }, /*run_in_background=*/false);
        EXPECT_EQ(handled, 1);
    });

    // サーバーの listen を待つ
    std::this_thread::sleep_for(50ms);

    RpcPipeClient cli{pipe_name};
    cli.connect(3000ms);

    Message resp = cli.send_request(Message{std::string_view{"ping"}}, 3000ms);
    EXPECT_EQ(resp.as_string_view(), "echo:ping");

    cli.close();
    ASSERT_NO_THROW(srv_fut.get());
}

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-03: 並行 send_request (3 スレッド) — 応答が混在しないこと
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcTest, SendRequest_MultipleParallel) {
    const auto pipe_name = unique_pipe("parallel");
    constexpr int N_THREADS = 3;
    constexpr int N_REQUESTS_PER_THREAD = 5;

    auto srv_fut = async_rpc_server(pipe_name, [](RpcPipeServer& srv) {
        // 背景スレッドで serve_requests
        srv.serve_requests([](const Message& req) -> Message {
            // 要求メッセージ "req:X:Y" → レスポンス "resp:X:Y"
            std::string s{req.as_string_view()};
            s.replace(0, 3, "resp");
            return Message{std::string_view{s}};
        }, /*run_in_background=*/true);

        // しばらく待ってから stop
        std::this_thread::sleep_for(3000ms);
        srv.stop();
    });

    std::this_thread::sleep_for(50ms);

    RpcPipeClient cli{pipe_name};
    cli.connect(3000ms);

    // 複数スレッドから並行して send_request
    std::vector<std::future<void>> futs;
    futs.reserve(N_THREADS);
    for (int t = 0; t < N_THREADS; ++t) {
        futs.push_back(std::async(std::launch::async, [&cli, t]() {
            for (int i = 0; i < N_REQUESTS_PER_THREAD; ++i) {
                std::string req_s = "req:" + std::to_string(t) + ":" + std::to_string(i);
                std::string exp_s = "resp:" + std::to_string(t) + ":" + std::to_string(i);
                Message resp = cli.send_request(Message{std::string_view{req_s}}, 3000ms);
                EXPECT_EQ(resp.as_string_view(), exp_s)
                    << "thread=" << t << " i=" << i;
            }
        }));
    }
    for (auto& f : futs) { ASSERT_NO_THROW(f.get()); }

    cli.close();
    ASSERT_NO_THROW(srv_fut.get());
}

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-04: タイムアウト → Timeout 例外
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcTest, SendRequest_Timeout_ThrowsTimeout) {
    const auto pipe_name = unique_pipe("timeout");

    auto srv_fut = async_rpc_server(pipe_name, [](RpcPipeServer& srv) {
        // リクエストを受け取っても応答しない（タイムアウトを誘発）
        std::this_thread::sleep_for(2000ms);
        srv.close();
    });

    std::this_thread::sleep_for(50ms);

    RpcPipeClient cli{pipe_name};
    cli.connect(3000ms);

    EXPECT_THROW({
        cli.send_request(Message{std::string_view{"no_reply"}}, 300ms);
    }, PipeException);

    cli.close();
    // サーバー側の close が先に来るので get() は無視
    try { srv_fut.get(); } catch (...) {}
}

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-05: 通常 send()/receive() と send_request() が共存できること
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcTest, PlainSend_CoexistsWithRpc) {
    const auto pipe_name = unique_pipe("coexist");

    auto srv_fut = async_rpc_server(pipe_name, [](RpcPipeServer& srv) {
        // (a) 通常フレームを受け取って返送
        Message plain = srv.receive(3000ms);
        srv.send(Message{plain});  // エコーバック

        // (b) RPC: 背景スレッドでサービス
        srv.serve_requests([](const Message& /*req*/) -> Message {
            return Message{std::string_view{"rpc_ok"}};
        }, /*run_in_background=*/true);

        std::this_thread::sleep_for(2000ms);
        srv.stop();
    });

    std::this_thread::sleep_for(50ms);

    RpcPipeClient cli{pipe_name};
    cli.connect(3000ms);

    // 通常 send→receive
    cli.send(Message{std::string_view{"plain_msg"}});
    Message plain_resp = cli.receive(3000ms);
    EXPECT_EQ(plain_resp.as_string_view(), "plain_msg");

    // RPC
    Message rpc_resp = cli.send_request(Message{std::string_view{"rpc_req"}}, 3000ms);
    EXPECT_EQ(rpc_resp.as_string_view(), "rpc_ok");

    cli.close();
    ASSERT_NO_THROW(srv_fut.get());
}

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-06: サーバー切断 → pending に ConnectionReset が届くこと
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcTest, ServerDisconnect_RejectsAllPending) {
    const auto pipe_name = unique_pipe("disconnect");

    auto srv_fut = async_rpc_server(pipe_name, [](RpcPipeServer& srv) {
        // 接続後 200ms で切断（応答しない）
        std::this_thread::sleep_for(200ms);
        srv.close();
    });

    std::this_thread::sleep_for(50ms);

    RpcPipeClient cli{pipe_name};
    cli.connect(3000ms);

    // send_request を投げて、サーバー切断で例外を受け取る
    EXPECT_THROW({
        cli.send_request(Message{std::string_view{"will_fail"}}, 5000ms);
    }, PipeException);

    cli.close();
    try { srv_fut.get(); } catch (...) {}
}

// ─────────────────────────────────────────────────────────────────────────────
// TC-F002-07: PROTOCOL_VERSION は 0x02
// ─────────────────────────────────────────────────────────────────────────────
TEST(RpcFrameTest, ProtocolVersion_Is_0x02) {
    EXPECT_EQ(detail::PROTOCOL_VERSION, uint8_t{0x02});
}

File: tests/python/test_rpc.py
# tests/python/test_rpc.py — RpcPipeClient / RpcPipeServer 統合テスト
# 仕様: spec/F002_rpc_message_id.md §10

import threading
import time
import pytest
import pipeutil
from conftest import unique_pipe


# ─── RPC サーバーヘルパー ─────────────────────────────────────────────────────

class RpcServerThread(threading.Thread):
    """
    背景スレッド上で RpcPipeServer を起動するヘルパー。

    usage:
        srv = RpcServerThread(pipe_name, handler=my_handler)
        srv.start()
        srv.wait_ready()    # accept 完了まで待機
        # ... クライアント操作 ...
        srv.shutdown()
        srv.join()
        srv.reraise()
    """

    def __init__(self, pipe_name: str, handler=None,
                 accept_timeout: float = 5.0,
                 serve_duration: float = 5.0):
        super().__init__(daemon=True)
        self.pipe_name = pipe_name
        self.handler = handler or (lambda req: req)  # デフォルト: エコー
        self.accept_timeout = accept_timeout
        self.serve_duration  = serve_duration
        self._ready = threading.Event()
        self._stop_ev = threading.Event()  # '_stop' は Python 3.8 Thread の内部メソッドと衝突するため回避
        self._exc: BaseException | None = None
        self._srv: pipeutil.RpcPipeServer | None = None

    def run(self):
        srv = pipeutil.RpcPipeServer(self.pipe_name)
        self._srv = srv
        try:
            srv.listen()
            self._ready.set()              # listen 後に set → クライアントが接続可能になる
            srv.accept(self.accept_timeout)
            srv.serve_requests(self.handler, run_in_background=True)
            # serve_duration 待ってから stop
            self._stop_ev.wait(self.serve_duration)
            srv.stop()
        except Exception as e:
            self._exc = e
            self._ready.set()  # エラーでもブロックしない
        finally:
            srv.close()

    def wait_ready(self, timeout: float = 5.0):
        assert self._ready.wait(timeout), "RpcServer did not start in time"

    def shutdown(self):
        self._stop_ev.set()

    def reraise(self):
        if self._exc:
            raise self._exc


# ─── テストクラス ─────────────────────────────────────────────────────────────

class TestRpc:

    # ─── TC-F002-PY-01: 単発 RPC ラウンドトリップ ─────────────────────────────
    def test_single_roundtrip(self):
        pipe_name = unique_pipe("rpc_single")

        def handler(req: pipeutil.Message) -> pipeutil.Message:
            return pipeutil.Message(b"pong:" + bytes(req))

        srv = RpcServerThread(pipe_name, handler=handler)
        srv.start()
        srv.wait_ready()

        cli = pipeutil.RpcPipeClient(pipe_name)
        try:
            cli.connect(3.0)
            resp = cli.send_request(pipeutil.Message(b"ping"), timeout=3.0)
            assert bytes(resp) == b"pong:ping"
        finally:
            cli.close()
            srv.shutdown()
            srv.join(timeout=5.0)
            srv.reraise()

    # ─── TC-F002-PY-02: 複数スレッドからの並行 send_request ──────────────────
    def test_concurrent_requests(self):
        pipe_name = unique_pipe("rpc_concurrent")
        N_THREADS  = 3
        N_REQUESTS = 5

        def handler(req: pipeutil.Message) -> pipeutil.Message:
            # "req:T:I" → "resp:T:I"
            s = bytes(req).decode()
            return pipeutil.Message(s.replace("req:", "resp:", 1).encode())

        srv = RpcServerThread(pipe_name, handler=handler, serve_duration=10.0)
        srv.start()
        srv.wait_ready()

        cli = pipeutil.RpcPipeClient(pipe_name)
        errors: list[str] = []
        lock = threading.Lock()

        def worker(t: int):
            for i in range(N_REQUESTS):
                req_b  = f"req:{t}:{i}".encode()
                exp_b  = f"resp:{t}:{i}".encode()
                resp = cli.send_request(pipeutil.Message(req_b), timeout=5.0)
                if bytes(resp) != exp_b:
                    with lock:
                        errors.append(
                            f"t={t} i={i}: got {bytes(resp)!r}, want {exp_b!r}")

        cli.connect(3.0)
        try:
            threads = [threading.Thread(target=worker, args=(t,)) for t in range(N_THREADS)]
            for th in threads: th.start()
            for th in threads: th.join(timeout=10.0)
        finally:
            cli.close()
            srv.shutdown()
            srv.join(timeout=5.0)
            srv.reraise()

        assert not errors, "\n".join(errors)

    # ─── TC-F002-PY-03: タイムアウト例外 ─────────────────────────────────────
    def test_timeout_raises(self):
        pipe_name = unique_pipe("rpc_timeout")

        # ハンドラが応答しない（time.sleep で遅延）
        def slow_handler(req: pipeutil.Message) -> pipeutil.Message:
            time.sleep(2.0)  # クライアントのタイムアウト (0.3s) より長いが合理的な範囲
            return pipeutil.Message(b"too_late")

        srv = RpcServerThread(pipe_name, handler=slow_handler, serve_duration=5.0)
        srv.start()
        srv.wait_ready()

        cli = pipeutil.RpcPipeClient(pipe_name)
        cli.connect(3.0)
        try:
            with pytest.raises(pipeutil.TimeoutError):
                cli.send_request(pipeutil.Message(b"fast_req"), timeout=0.3)
        finally:
            cli.close()
            srv.shutdown()
            srv.join(timeout=5.0)
            # サーバー側の例外は無視（slow_handler がブロックしている可能性）
